PHOENIX-2628 Ensure split when iterating through results handled corr…

chrajeshbabu opened the pull request:
The patch fixes issues with splits and merges while scanning local indexes.

JamesRTaylor commented on phoenix-core/src/it/java/org/apache/phoenix/end2end/index/LocalIndexIT.java:
```diff
@@ -727,7 +727,7 @@ public void testLocalIndexScanAfterRegionSplit() throws Exception {
             
             HBaseAdmin admin = driver.getConnectionQueryServices(getUrl(), TestUtil.TEST_PROPERTIES).getAdmin();
-            for (int i = 1; i < 5; i++) {
+            for (int i = 1; i < 2; i++) {
```
Any reason to go from 5 to 2 here? Maybe get rid of the loop then? Does it decrease the amount of testing being done?

chrajeshbabu replied:
While debugging a failure changed it to 2. 5 is ok will keep it as it is.

JamesRTaylor commented on phoenix-core/src/it/java/org/apache/phoenix/end2end/index/MutableIndexIT.java:
```diff
@@ -86,8 +98,8 @@ public static void doSetup() throws Exception {
 	@Parameters(name="localIndex = {0} , transactional = {1}")
     public static Collection<Boolean[]> data() {
-        return Arrays.asList(new Boolean[][] {     
-                 { false, false }, { false, true }, { true, false }, { true, true }
```
Whitespace only change?

chrajeshbabu replied:
Yes James it's space only change.

JamesRTaylor commented on phoenix-core/src/main/java/org/apache/hadoop/hbase/regionserver/IndexHalfStoreFileReaderGenerator.java:
```diff
@@ -157,6 +162,7 @@ public Reader preStoreFileReaderOpen(ObserverContext<RegionCoprocessorEnvironmen
             } catch (ClassNotFoundException e) {
                 throw new IOException(e);
             } catch (SQLException e) {
+                e.printStackTrace();
```
Remove as we're throwing here

chrajeshbabu replied:
Yes. This can removed.

JamesRTaylor commented on phoenix-core/src/main/java/org/apache/phoenix/compile/ScanRanges.java:
```diff
@@ -385,9 +387,25 @@ public Scan intersectScan(Scan scan, final byte[] originalStartKey, final byte[]
-        newScan.setAttribute(SCAN_ACTUAL_START_ROW, scanStartKey);
-        newScan.setStartRow(scanStartKey);
-        newScan.setStopRow(scanStopKey);
+        if(ScanUtil.isLocalIndex(scan)) {
```
The code in this if statement can go in the if (keyOffset > 0) if statement instead. We're trying to avoid checks for local indexes.

chrajeshbabu replied:
Sure will try to remove the local index scan check.

chrajeshbabu replied:
@JamesRTaylor  we cannot use keyoffset > 0 check always for considering local indexes scan because in special case like if a table has only one region then region start key length and end key length is zero so keyoffset also becoming zero. In that case we need to check local index scan or not to set the attributes properly. Wdyt?

JamesRTaylor replied:
How about moving the setting of these special attributes outside of this method?

chrajeshbabu replied:
I will try moving outside of this method.

JamesRTaylor commented on phoenix-core/src/main/java/org/apache/phoenix/coprocessor/BaseScannerRegionObserver.java:
```diff
@@ -165,7 +183,7 @@ public RegionScanner preScannerOpen(final ObserverContext<RegionCoprocessorEnvir
         }
         if (isRegionObserverFor(scan)) {
-            if (! skipRegionBoundaryCheck(scan)) {
+            if (! skipRegionBoundaryCheck(scan) || ScanUtil.isLocalIndex(scan)) {
```
Add comment, please. Something like this:

    // For local indexes, we need to throw if out of region as we'll get duplicates otherwise while in other cases,
    // it may just mean out client-side data on region boundaries is out of date and can safely be ignored.

Note that we'll need to handle this from the MR integration which might be painful.

JamesRTaylor commented on phoenix-core/src/main/java/org/apache/phoenix/coprocessor/BaseScannerRegionObserver.java:
```diff
@@ -337,6 +384,22 @@ public boolean nextRaw(List<Cell> result) throws IOException {
+                            }
+                        }
+                        System.out.println("&&&&-< "+scan);
+                        System.out.println("&&&&-<< "+result);
```
Remove System.out.println calls

JamesRTaylor commented on phoenix-core/src/main/java/org/apache/phoenix/coprocessor/BaseScannerRegionObserver.java:
```diff
@@ -279,6 +301,31 @@ protected RegionScanner getWrappedScanner(final ObserverContext<RegionCoprocesso
+            private HRegionInfo regionInfo = c.getEnvironment().getRegionInfo();
+            private byte[] actualStartKey = getActualStartKey();
+
+            private boolean checkForReferenceFiles(){
```
Please add a code comment about why we need to check for references. Do we always have to do this? Is it expensive?

chrajeshbabu replied:
Added the code comments and checking references is not expensive. We need to do this for local indexes case only.

JamesRTaylor commented on phoenix-core/src/main/java/org/apache/phoenix/coprocessor/BaseScannerRegionObserver.java:
```diff
@@ -279,6 +301,31 @@ protected RegionScanner getWrappedScanner(final ObserverContext<RegionCoprocesso
+            }
+
             @Override
             public boolean next(List<Cell> results) throws IOException {
```
Can we refactor the code between these two next(List<Cell>, ...) implementations, as there's now a fair amount of code which is duplicated here?

JamesRTaylor commented on phoenix-core/src/main/java/org/apache/phoenix/coprocessor/BaseScannerRegionObserver.java:
```diff
@@ -337,6 +384,22 @@ public boolean nextRaw(List<Cell> result) throws IOException {
                         arrayElementCell = result.get(arrayElementCellPosition);
                     }
                     if (ScanUtil.isLocalIndex(scan) && !ScanUtil.isAnalyzeTable(scan)) {
```
We'll need some code comments here. Are we skipping until we find the actualStartKey row? Would it be better/possible to do a seek? This seems like it'd be potentially expensive, no?

chrajeshbabu replied:
We are getting the results after seek only. Will check once again whether we can handle this in local index scanner.

JamesRTaylor commented on phoenix-core/src/main/java/org/apache/phoenix/coprocessor/GroupedAggregateRegionObserver.java:
```diff
@@ -423,7 +426,14 @@ private RegionScanner scanUnordered(ObserverContext<RegionCoprocessorEnvironment
+                    Exception cause = new StaleRegionBoundaryCacheException(c.getEnvironment().getRegion().getRegionInfo().getTable().getNameAsString());
+                    throw new DoNotRetryIOException(cause.getMessage(), cause);
+                } else {
+                    throw e;
```
Is the throw ok because the client scanner knows how to deal with this in cases other than local indexes? Let's doc that, please.

chrajeshbabu replied:
There is a chance of split can happen after preScannerOpen and before postScannerOpen. I think we can throw stale region boundary exception all the cases because even when we throw NSRE we again need to open the scanner which checks the region boundaries and throw the stale region boundary exception any way. Will check it once.

JamesRTaylor commented on phoenix-core/src/main/java/org/apache/phoenix/coprocessor/GroupedAggregateRegionObserver.java:
```diff
@@ -402,8 +405,8 @@ private RegionScanner scanUnordered(ObserverContext<RegionCoprocessorEnvironment
             Region region = c.getEnvironment().getRegion();
-            region.startRegionOperation();
             try {
+                region.startRegionOperation();
```
Doesn't startRegionOperation() take a lock which would prevent a split from occurring and thus a NotServingRegionException from happening? Or is this the case where the split occurred after pre open, but before we start iterating? Let's doc that please.

chrajeshbabu replied:
region.startRegionOperation() prevent from splitting and also if the region already closed we are throwing NSRE. I have moved into try block only to catch NSRE and throw Stale region boundary exception.

chrajeshbabu replied:
Currently changed the code to throw stale region exception all when we get NSRE because hbase client may recreate the scanners with wrong region boundaries on NSRE. I have handling this in BaseScannerRegionObserver so need not handle in all observers.

JamesRTaylor commented on phoenix-core/src/main/java/org/apache/phoenix/execute/BaseQueryPlan.java:
```diff
@@ -196,26 +196,31 @@ public Expression getDynamicFilter() {
         // Set miscellaneous scan attributes. This is the last chance to set them before we
         // clone the scan for each parallelized chunk.
-        Scan scan = context.getScan();
+        Scan scan = s == null ? context.getScan() : s;
```
Is this null check required?

JamesRTaylor commented on phoenix-core/src/main/java/org/apache/phoenix/execute/ScanPlan.java:
```diff
@@ -209,7 +209,11 @@ protected ResultIterator newIterator(ParallelScanGrouper scanGrouper) throws SQL
+                    && ScanUtil.shouldRowsBeInRowKeyOrder(orderBy, context)
+                    || (table.getIndexType() == IndexType.LOCAL && (Bytes.compareTo(
+                        scan.getStartRow(), this.context.getScan().getStartRow()) != 0 || Bytes
+                            .compareTo(scan.getStopRow(), this.context.getScan().getStopRow()) != 0))) {
```
Why is  a merge sort required for this new condition? Is this the case when the original start/stop row has changed? Is this purely for the chunked result iterator case? Let's document this and add a TODO to remove as the chunked result iterator is deprecated (and if you could consistently add this TODO for other parts of code only necessary due to this deprecated chunking mechanism, that'd be much appreciated).

JamesRTaylor commented on phoenix-core/src/main/java/org/apache/phoenix/iterate/BaseResultIterators.java:
```diff
@@ -326,11 +325,12 @@ private static void optimizeProjection(StatementContext context, Scan scan, PTab
         this.plan = plan;
         this.scanGrouper = scanGrouper;
         StatementContext context = plan.getContext();
+        this.scan = scan == null ? context.getScan() : scan;
```
When is scan null?

chrajeshbabu replied:
This null check is not required will remove it.

JamesRTaylor commented on phoenix-core/src/main/java/org/apache/phoenix/iterate/BaseResultIterators.java:
```diff
@@ -465,7 +465,14 @@ private static String toString(List<byte[]> gps) {
+        if (scan == null
+                || (ScanUtil.isLocalIndex(scan)
+                        && Bytes.compareTo(context.getScan().getStartRow(), scan.getStartRow()) == 0 && Bytes
+                        .compareTo(context.getScan().getStopRow(), scan.getStopRow()) == 0)) {
```
Cover with ScanUtil method, as I've seen this code repeated a number of times.

chrajeshbabu replied:
Sure will move to ScanUtil

JamesRTaylor commented on phoenix-core/src/main/java/org/apache/phoenix/iterate/BaseResultIterators.java:
```diff
@@ -465,7 +465,14 @@ private static String toString(List<byte[]> gps) {
+                || (ScanUtil.isLocalIndex(scan)
+                        && Bytes.compareTo(context.getScan().getStartRow(), scan.getStartRow()) == 0 && Bytes
+                        .compareTo(context.getScan().getStopRow(), scan.getStopRow()) == 0)) {
+            return getParallelScans(EMPTY_BYTE_ARRAY, EMPTY_BYTE_ARRAY);
```
This seems strange - why do we call getParallelScans with empty byte arrays in this case? Please document.

chrajeshbabu replied:
This check detects whether scan boundaries  are equal to context scan boundaries or not. If they are same we are going by getting all parallel scans for the table. Will document it.

JamesRTaylor commented on phoenix-core/src/main/java/org/apache/phoenix/iterate/BaseResultIterators.java:
```diff
@@ -556,35 +564,55 @@ private static String toString(List<byte[]> gps) {
-                scans = addNewScan(parallelScans, scans, newScan, endKey, true, regionLocation);
-                currentKeyBytes = endKey;
+                if (Bytes.compareTo(scan.getStartRow(), context.getScan().getStartRow()) != 0
+                     || Bytes.compareTo(scan.getStopRow(), context.getScan().getStopRow()) != 0) {
```
Cover with ScanUtil function

JamesRTaylor replied:
Also, this code path is strange. What's the entry point? Does it need to be mixed up with the existing code? It feels like we've completely forked the flow here. Why not have it in it's own method called only when this special case is detected?

chrajeshbabu replied:
If I move creating parallel scans for this special case I need to duplicate lot of code. That's why I have added special case as part of creating existing parallel scans only.

JamesRTaylor replied:
Seems like you've forked the bulk of the code completely. The only think you'd duplicate would be the loop and a bit of setup code up top (that could be factored out). What's the entry point for this? Is the problem that the entry point is the same?

chrajeshbabu replied:
getParalleScans is the entry point. Let me move it out.

JamesRTaylor commented on phoenix-core/src/main/java/org/apache/phoenix/iterate/BaseResultIterators.java:
```diff
@@ -556,35 +564,55 @@ private static String toString(List<byte[]> gps) {
+                    Scan newScan = ScanUtil.newScan(scan);
+                    if(ScanUtil.isLocalIndex(scan)) {
+                        newScan.setStartRow(regionInfo.getStartKey());
+                        newScan.setAttribute(SCAN_ACTUAL_START_ROW, regionInfo.getStartKey());
```
Is there some symmetry between SCAN_ACTUAL_START_ROW and EXPECTED_UPPER_REGION_KEY? Should they be named similarly?

chrajeshbabu replied:
First one is actual start key of the scan second one is end key of region. After this patch we don't need EXPECTED_UPPER_REGION_KEY but kept it for compatibility.

JamesRTaylor replied:
If we don't need EXPECTED_UPPER_REGION_KEY, let's remove it.

JamesRTaylor commented on phoenix-core/src/main/java/org/apache/phoenix/iterate/ChunkedResultIterator.java:
```diff
@@ -56,6 +57,7 @@
     private PeekingResultIterator resultIterator;
+    private QueryPlan plan;
 
     public static class ChunkedResultIteratorFactory implements ParallelIteratorFactory {
```
FYI, this class is deprecated now with newer versions of HBase.

chrajeshbabu replied:
Currently removed the changes required for ChunkedResultIterator. The changes in the chunked result iterator are just removing code not required.

JamesRTaylor commented on phoenix-core/src/main/java/org/apache/phoenix/iterate/TableResultIterator.java:
```diff
@@ -80,13 +94,19 @@
         this.scanIterator = UNINITIALIZED_SCANNER;
         this.renewLeaseThreshold = renewLeaseThreshold;
+        this.plan = plan;
+        this.handleSplitRegionBoundaryFailureDuringInitialization = handleSplitRegionBoundaryFailureDuringInitialization;
```
When is handleSplitRegionBoundaryFailureDuringInitialization true versus false? Why would we *not* want to handle it?

chrajeshbabu replied:
handleSplitRegionBoundaryFailureDuringInitialization will be true for scanning intermediate chunks in ChunkedResultIterator. When we create scanner for new chunk we might see stale region boundaries in that case also we need to forcibly recreate the iterator with new boundaries. In normal case if we get StaleRegionBoundary exception while creating scanner BaseResultIterator handles it.

JamesRTaylor commented on phoenix-core/src/main/java/org/apache/phoenix/iterate/TableResultIterator.java:
```diff
@@ -107,8 +127,37 @@ public synchronized void close() throws SQLException {
+                        if(ScanUtil.isLocalIndex(newScan)) {
+                            newScan.setAttribute(SCAN_START_ROW_SUFFIX, ByteUtil.nextKey(startRowSuffix));
+                        } else {
+                            newScan.setStartRow(ByteUtil.nextKey(startRowSuffix));
```
I'm not convinced that this nextKey will universally work. Probably better (if we really want to just get the next possible key) to add a 0 byte instead.

chrajeshbabu replied:
Will raise improvement action for this James.

JamesRTaylor replied:
See other comment below, but how about we serialize the row key we found which is out-of-region in the StaleRegionBoundaryCacheException instead of mucking with the startRowSuffix?

JamesRTaylor commented on phoenix-core/src/main/java/org/apache/phoenix/iterate/TableResultIterator.java:
```diff
@@ -107,8 +127,37 @@ public synchronized void close() throws SQLException {
+                    }
+                    plan.getContext().getConnection().getQueryServices().clearTableRegionCache(htable.getTableName());
+                    this.scanIterator =
+                            plan.iterator(DefaultParallelScanGrouper.getInstance(), newScan);
```
Getting the ParallelScanGrouper here as state instead of assuming the default one is being used may be necessary (or making it not necessary to have this ParallelScanGrouper - it's an edge case for the MR integration - there's likely a better way - perhaps the plan itself could know what to use for the ParallelScanGrouper?)

JamesRTaylor replied:
So we're re-running the scan, only if it's an aggregate scan? Does this apply only for an "ordered" group by (which is grouping by the leading part of the pk and doing so in-place)? In the unordered case, each scan is done during the post open and the results are cached on the RS, so a split won't impact the results. There may be a race condition between the pre open (where we check if we're in the region) and the post open (where we start the region operation), but I think that could be solved by doing the in-region check again). In these cases, we already handle this through the exception catching code in BaseResultIterators.

chrajeshbabu replied:
Yes aggregate queries are already handled properly. This code is only for handling splits when we are in the middle of non aggregate queries. If there are splits in the starting of the query then we are throwing out the stale region exception to BaseResultIterators which handles creating the proper parallel scans.
+                if(scan.getAttribute(NON_AGGREGATE_QUERY)!=null) {

JamesRTaylor replied:
Ah, I misread that. The case of an ordered aggregate falls into the same category, however, as we're aggregating in-place. I don't think we need any additional code in the unordered aggregate code path in GroupedAggregateRegionObserver (other than the duplicate in-region check I mentioned above).

chrajeshbabu replied:
You mean we can check the in region boundary again in post scanner open and if it's out of range throw stale region boundary exception which will be handled by BaseResultIterators?

JamesRTaylor replied:
Yes, exactly. Except for the ordered aggregation case, which needs to be handled as the non aggregate is handled (i.e. we'd still throw the stale region boundary exception, but the client would need to react to it as you've done for the non aggregate case. We need to know which exact row key we're on though (which the client won't know). Seems like the more feasible approach would be to include the row key in the exception and deserialize it on the client side.

chrajeshbabu replied:
Agree with you James. I think I can raise another issue for this and work on it. Wdyt?

JamesRTaylor commented on phoenix-core/src/main/java/org/apache/phoenix/iterate/TableResultIterator.java:
```diff
@@ -107,8 +127,37 @@ public synchronized void close() throws SQLException {
+            try {
+                throw ServerUtil.parseServerException(e);
+            } catch(StaleRegionBoundaryCacheException e1) {
+                if(scan.getAttribute(NON_AGGREGATE_QUERY)!=null) {
```
Minor nit: how about adding a ScanUtil.isAggregateQuery() method instead?

chrajeshbabu replied:
Will move the check to ScanUtil.

JamesRTaylor commented on phoenix-core/src/main/java/org/apache/phoenix/iterate/TableResultIterator.java:
```diff
@@ -107,8 +127,37 @@ public synchronized void close() throws SQLException {
+        } catch (SQLException e) {
+            try {
+                throw ServerUtil.parseServerException(e);
+            } catch(StaleRegionBoundaryCacheException e1) {
```
This needs code comments, please.

chrajeshbabu replied:
I will add.

JamesRTaylor commented on phoenix-core/src/main/java/org/apache/phoenix/iterate/TableResultIterator.java:
```diff
@@ -121,8 +170,21 @@ public synchronized void initScanner() throws SQLException {
             } catch (IOException e) {
-                Closeables.closeQuietly(htable);
-                throw ServerUtil.parseServerException(e);
+                if(handleSplitRegionBoundaryFailureDuringInitialization) {
```
Rather than have a handleSplitRegionBoundaryFailureDuringInitialization, would it work to always throw it and let the caller deal with it? When a variable name gets beyond 50 characters, it's a sign that it may be too special of a case. :-)

chrajeshbabu replied:
This is required for ChunkedResultIterator if we are going to deprecate or not use it we don't need these changes. Other than that we might need to deal with it in PhoenixRecordReader(currently we skipping range check but for local indexes it's compulsory to handle otherwise we might miss one of the daughter records)

JamesRTaylor commented:
Thanks for the patch, @chrajeshbabu. This is a big improvement over your initial approach. It's still complex, though. There's an on going cost to complicating the code to this degree that would be good to minimize. 

One big thing missing is an overview of all the moving parts. Perhaps that could live in BaseResultIterators? For example - this exception is thrown under these conditions and this state is added to the scan to know how to adjust the scan ranges, etc. etc. Especially that hasReferences bit in the coprocessor - what's that all about?

I'd like to understand how ChunkedResultIterator complicates things too. I think it'd be ok to document that splits during aggregate queries aren't handled if your HBase version is less that XXX if it'd significantly simplify this patch (as that code is essentially deprecated). 

Another potential, different approach would be for Phoenix to universally handle the split during scan case (rather than letting the HBase client scanner handle it for non aggregate case and Phoenix handle it for the aggregate case). Would that simplify things?

JamesRTaylor commented on phoenix-core/src/main/java/org/apache/phoenix/iterate/BaseResultIterators.java:
```diff
@@ -556,35 +564,55 @@ private static String toString(List<byte[]> gps) {
+                        }
+                    }   
+                    scans = addNewScan(parallelScans, scans, newScan, endKey, true, regionLocation);
+                 } else {
```
Can we remove the if (isLocalIndex) checks below since the above if statement would have been entered instead?

chrajeshbabu commented:
@JamesRTaylor  Thanks for the review. Committed the changes handling the review comments.
Refactored the code and added code comments where ever possible. 

Now not handling the special cases required for ChunkedResultItertor. 

bq. Another potential, different approach would be for Phoenix to universally handle the split during scan case (rather than letting the HBase client scanner handle it for non aggregate case and Phoenix handle it for the aggregate case). Would that simplify things?
Now throwing stale region bound exception all the cases when ever we get NSRE then phoenix client can handle the NSRE than hbase client handling with wrong region boundaries.

The ordered aggregate queries issue you are mentioning can be handled as part of other issue.

Please review the latest code.
Thanks.

JamesRTaylor commented:
+1. Looks great, @chrajeshbabu. Thanks for the excellent work.

chrajeshbabu commented:
Thanks @JamesRTaylor  for review. Will rebase the patch and commit.
